### PR TITLE
implement lindi.File, drop-in replacement for h5py.File that supports .lindi.json files based on extension (4)

### PR DESCRIPTION
**Pull request by magland**

### CODE REVIEW

**Comment by magland** [comment]
Here we make a lindi.File that can be a drop-in replacement for h5py.File, where it smartly decides whether to use h5py or lindi based on the file extension of the file name.

This PR corrects a few small issues/mistakes from previous PRs.

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 1]
```diff
@@ -52,6 +52,9 @@ def from_lindi_file(url_or_path: str, *, mode: LindiFileMode = "r", staging_area
 
         For a description of parameters, see from_reference_file_system().
         """
+        if local_file_path is None:
+            if not url_or_path.startswith("http://") and not url_or_path.startswith("https://"):
+                local_file_path = url_or_path
```
Comment (magland) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:57]: Set the local file path (the one that can get edited) to be equal to the url_or_path in the case that it is not provided and is not a url

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 2]
```diff
@@ -166,7 +169,7 @@ def from_reference_file_system(rfs: Union[dict, str, None], *, mode: LindiFileMo
             store = LindiReferenceFileSystemStore(rfs, local_cache=local_cache)
             if staging_area:
                 store = LindiStagingStore(base_store=store, staging_area=staging_area)
-            return LindiH5pyFile.from_zarr_store(store, mode=mode, local_file_path=local_file_path)
+            return LindiH5pyFile.from_zarr_store(store, mode=mode, local_file_path=local_file_path, local_cache=local_cache)
```
Comment (magland) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:172]: This was accidentally omitted.

File: lindi/LindiStagingStore/StagingArea.py [hunk 3]
```diff
@@ -21,17 +22,27 @@ def __init__(self, *, _directory: str) -> None:
...
+            If provided, the exact directory where the staging area will be
+            created. It is okay if this directory already exists.
         """
-        dir = os.path.join(base_dir, _create_random_id())
+        if base_dir is not None and dir is not None:
+            raise ValueError("Provide either base_dir or dir, but not both")
+        if base_dir is not None:
+            dir = os.path.join(base_dir, _create_random_id())
+        if dir is None:
+            raise ValueError("Provide either base_dir or dir")
```
Comment (magland) [inline lindi/LindiStagingStore/StagingArea.py:45]: Either supply a dir or a base_dir

File: .gitignore [hunk 4]
```diff
@@ -1,4 +1,4 @@
-*.lindi.json
+*.lindi.json*
```
Comment (rly) [inline .gitignore:1]: Why this change?

Comment (rly) [reply .gitignore:1]: Never mind. I see that the staging directories take the file name and add ".d"

**Review by rly — APPROVED** [review]
